azblob/zt_client_provided_key_test.go

```diff
@@ -0,0 +1,504 @@
+package azblob
+
+import (
+	"bytes"
+	"context"
+	"crypto/md5"
+	"encoding/base64"
+	"encoding/binary"
+	"io/ioutil"
+	"strconv"
+	"strings"
+	"time"
+
+	chk "gopkg.in/check.v1" // go get gopkg.in/check.v1
+)
+
+/*
+Azure Storage supports following operations support of sending customer-provided encryption keys on a request:
+Put Blob, Put Block List, Put Block, Put Block from URL, Put Page, Put Page from URL, Append Block,
+Set Blob Properties, Set Blob Metadata, Get Blob, Get Blob Properties, Get Blob Metadata, Snapshot Blob.
+*/
+var testEncryptedKey = "MDEyMzQ1NjcwMTIzNDU2NzAxMjM0NTY3MDEyMzQ1Njc="
+var testEncryptedHash = "3QFFFpRA5+XANHqwwbT4yXDmrT/2JaLt/FKHjzhOdoE="
+var testEncryptedScope = ""
+var testCPK = InitClientProvidedKeyOptions(&testEncryptedKey, &testEncryptedHash, &testEncryptedScope)
+
+func blockIDBinaryToBase64(blockID []byte) string {
+	return base64.StdEncoding.EncodeToString(blockID)
+}
+
+func blockIDBase64ToBinary(blockID string) []byte {
+	binary, _ := base64.StdEncoding.DecodeString(blockID)
+	return binary
+}
+
+// blockIDIntToBase64 functions convert an int block ID to a base-64 string and vice versa
+func blockIDIntToBase64(blockID int) string {
+	binaryBlockID := (&[4]byte{})[:] // All block IDs are 4 bytes long
+	binary.LittleEndian.PutUint32(binaryBlockID, uint32(blockID))
+	return blockIDBinaryToBase64(binaryBlockID)
+}
+func blockIDBase64ToInt(blockID string) int {
+	blockIDBase64ToBinary(blockID)
+	return int(binary.LittleEndian.Uint32(blockIDBase64ToBinary(blockID)))
+}
+
+func (s *aztestsSuite) TestPutBlockAndPutBlockListWithCPK(c *chk.C) {
+	bsu := getBSU()
+	container, _ := createNewContainer(c, bsu)
+	defer delContainer(c, container)
+
+	blobURL := container.NewBlockBlobURL(generateBlobName())
+
+	words := []string{"AAA ", "BBB ", "CCC "}
+	base64BlockIDs := make([]string, len(words))
+	for index, word := range words {
+		base64BlockIDs[index] = blockIDIntToBase64(index)
+		_, err := blobURL.StageBlock(ctx, base64BlockIDs[index], strings.NewReader(word), LeaseAccessConditions{}, nil, *testCPK)
+		if err != nil {
+			c.Fail()
+		}
+	}
+
+	resp, err := blobURL.CommitBlockList(ctx, base64BlockIDs, BlobHTTPHeaders{}, Metadata{}, BlobAccessConditions{}, *testCPK)
+	if err != nil {
+		c.Fail()
+	}
+
+	c.Assert(resp.ETag(), chk.NotNil)
+	c.Assert(resp.LastModified(), chk.NotNil)
+	c.Assert(resp.IsServerEncrypted(), chk.Equals, "true")
+	c.Assert(resp.EncryptionKeySha256(), chk.DeepEquals, *(testCPK.EncryptionKeySha256))
+
+	// Get blob content without encryption key should fail the request.
+	_, err = blobURL.Download(ctx, 0, 0, BlobAccessConditions{}, false, ClientProvidedKeyOptions{})
+	if err == nil {
+		c.Fail()
+	}
+
+	// Download blob to do data integrity check.
+	getResp, err := blobURL.Download(ctx, 0, 0, BlobAccessConditions{}, false, *testCPK)
+	if err != nil {
+		c.Fail()
+	}
```

Comment:
same here, we can use chk.IsNil which is one line

Reply:
Done